internal/controllers/core/podlogstream/podsource.go

```diff
@@ -72,45 +80,101 @@ func (s *PodSource) TearDown() {
 // Register the pods for this stream.
 //
 // Set up any watches we need.
-func (s *PodSource) handleReconcileRequest(ctx context.Context, name types.NamespacedName, pls *PodLogStream) {
+func (s *PodSource) handleReconcileRequest(plsNN types.NamespacedName, pls *PodLogStream) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.indexer.OnReconcile(name, pls)
+	s.indexer.OnReconcile(plsNN, pls)
 
-	ns := pls.Spec.Namespace
+	var ns string
+	if pls != nil {
+		ns = pls.Spec.Namespace
+	}
+
+	// ensure namespace watch registered if PLS object exists + has valid spec
 	if ns != "" {
-		_, ok := s.watchesByNamespace[ns]
+		pw, ok := s.watchesByNamespace[ns]
 		if !ok {
-			ctx, cancel := context.WithCancel(ctx)
-			pw := podWatch{ctx: ctx, cancel: cancel, namespace: ns}
+			pw = newPodWatch(s.ctx, ns)
 			s.watchesByNamespace[ns] = pw
-			go s.doWatch(pw)
+			go s.doWatchWithRetry(pw)
+		}
+		pw.watchers[plsNN] = true
+	}
+
+	// clean up any pre-existing namespace watches
+	// this covers both PLS changing namespace in spec + PLS deletion
+	for nsKey, watch := range s.watchesByNamespace {
+		if watch.watchers[plsNN] && nsKey != ns {
+			delete(watch.watchers, plsNN)
+			s.maybeCleanupWatch(nsKey)
 		}
 	}
 }
 
-// Process pod events and make sure they trigger a reconcile.
-func (s *PodSource) doWatch(pw podWatch) {
-	defer pw.cancel()
+// maybeCleanupWatch removes a namespace watch if there are no more active watchers.
+func (s *PodSource) maybeCleanupWatch(nsKey string) {
+	watch, ok := s.watchesByNamespace[nsKey]
+	if !ok {
+		return
+	}
 
-	podCh, err := s.kClient.WatchPods(s.ctx, k8s.Namespace(pw.namespace))
-	if err != nil {
-		logger.Get(pw.ctx).Errorf("watching pods: %v", err)
+	if len(watch.watchers) != 0 {
+		// there's still active watchers
 		return
 	}
 
+	watch.cancel()
+	delete(s.watchesByNamespace, nsKey)
+}
+
+func (s *PodSource) doWatchWithRetry(pw podWatch) {
```

Comment:
oof, how far are you down this approach?

i'm not a fan of having an internal restart loop like this, rather than letting things bubble up to the main reconciler and let the reconciler handle the retry (so that it can also publish the error state to the api)

i have a branch that i think might handle this (in the spirit of the other controller-cleanup prs)

Reply:
I can back this out! I'll pause on PLS changes for now while you get those changes in.